File: 알고리즘/Graph/1504.js
let fs = require("fs");
// let input = fs.readFileSync("/dev/stdin").toString().trim().split("\n")
let input = fs.readFileSync("test.txt").toString().trim().split("\n");

const [vertex, edge] = input[0].split(" ").map((num) => parseInt(num));

function Graph() {
    this.edges = {};
}

Graph.prototype.addVertex = function (v) {
    this.edges[v] = {};
};

Graph.prototype.addEdge = function (v1, v2, w) {
    this.edges[v1][v2] = w;
    this.edges[v2][v1] = w;
};

const graph = new Graph();

for (let i = 1; i <= vertex; i++) {
    graph.addVertex(i);
}

for (let i = 1; i < input.length - 1; i++) {
    const [v1, v2, w] = input[i].split(" ").map((num) => parseInt(num));
    graph.addEdge(v1, v2, w);
}

const [v1, v2] = input[input.length-1].split(" ").map((num) => parseInt(num))

function PrioQ() {
    this.heap = [];
}

PrioQ.prototype.swap = function(id, pid) {
    const temp = this.heap[pid];
    this.heap[pid] = this.heap[id];
    this.heap[id] = temp;
}

PrioQ.prototype.getPid = function(id) {
    return Math.floor((id-1) / 2)
}

PrioQ.prototype.pWeight = function(id) {
    return this.heap[this.getPid(id)]
}

PrioQ.prototype.heappush = function (vertex, weight) {
    this.heap.push([vertex, weight]);
    let id = this.heap.length - 1;
    let pid = this.getPid(id);
    while (this.heap[pid] && this.pWeight(id) > weight) {
        this.swap(id, pid);
        id = pid;
        pid = this.getPid(id);
    }
}

PrioQ.prototype.heappop = function () {
    if (this.heap.length === 1) {
        return this.heap.pop();
    }
    const result = this.heap[0];
    this.swap(0, this.heap.length -1);
    this.heap.pop();
    let id = 0;
    const weight = this.heap[0][1];
    let cid;
    while ((this.leftChild(id) && this.leftChild(id)[1] < weight) || (this.rightChild(id) && this.rightChild(id)[1] < weight)) {
        cid = this.leftChildIndex(id);
        if (this.rightChild(id) && this.rightChild(id)[1] < this.leftChild(id)[1]) {
            cid = this.rightChildIndex(id);
        }
        this.swap(id, cid);
        id = cid;
    }
    return result;
}

PrioQ.prototype.leftChild = function(id) {
    const cid = this.leftChildIndex(id);
    return this.heap[cid];
}

PrioQ.prototype.rightChild = function(id) {
    const cid = this.rightChildIndex(id);
    return this.heap[cid];
}

PrioQ.prototype.leftChildIndex = function(id) {
    return 2*id + 1;
}

PrioQ.prototype.rightChildIndex = function(id) {
    return 2*id + 2;
}

Graph.prototype.dijkstra = function (start) {
    const queue = new PrioQ();
    const distance = {};
    const visited = {};
    let total = graph.edges[v1][v2];
    let preVertex = start;
    for(let vertex in graph.edges) {
        distance[vertex] = Number.POSITIVE_INFINITY;
        visited[vertex] = false;
    }
    visited[start] = true;
    distance[start] = 0;
    distance[v2] = graph.edges[v1][v2]
    for (let [vertex, weight] of Object.entries(graph.edges[start])) {
        queue.heappush(vertex, weight)
    }
    while (queue.heap.length !== 0) {
        const [minVertex, minWeight] = queue.heappop();
        if (!visited[minVertex]) {
            visited[minVertex] = true;
            
            for (let [vertex, weight] of Object.entries(graph.edges[minVertex])) {
                queue.heappush(vertex, weight + minWeight)
            }
            if (parseInt(minVertex) !== v2 && distance[minVertex] > minWeight) {
                console.log(`from ${preVertex} to ${minVertex}`)
                total += graph.edges[preVertex][minVertex]
                distance[minVertex] = minWeight;
            }
            preVertex = minVertex;
        }
    }

    if (Object.values(distance).includes(Number.POSITIVE_INFINITY)) return -1;
    return total
}
console.log(graph.dijkstra(v1))

File: 알고리즘/Graph/1504.py
import heapq
import sys
input = sys.stdin.readline

class Graph:
    def __init__(self):
        self.items = {}

    def addVertex (self, v):
        self.items[v] = {}
    
    def addEdge (self,v1,v2,w):
        if self.items[v1].get(v2):
            self.items[v1][v2] = w if w < self.items[v1][v2] else self.items[v1][v2]
        else:
            self.items[v1][v2] = w
    
    def dijkstra(self, start, target, N):
        distance = {}
        edges = {}
        heap = [[0, start]]
        for i in range(1,N+1):
            distance[i] = float("inf")
            edges[i] = self.items[i].copy()
        distance[start] = 0
        while heap:
            dist, node = heapq.heappop(heap)
            if edges.get(node):
                for n_node, weight in edges[node].items():
                    if distance[n_node] > dist + weight:
                        distance[n_node] = dist + weight
                        heapq.heappush(heap,[distance[n_node], n_node])
                del edges[node]

        return distance[target]

n,e = map(int,input().split())
graph = Graph()
for i in range(1,n+1):
    graph.addVertex(i)

for _ in range(e):
    v1,v2,w = map(int,input().split())
    graph.addEdge(v1,v2,w)
    graph.addEdge(v2,v1,w)

v1,v2 = map(int,input().split())
answer = min(graph.dijkstra(1,v1,n) + graph.dijkstra(v1,v2,n) + graph.dijkstra(v2,n,n),graph.dijkstra(1,v2,n) + graph.dijkstra(v2,v1,n) + graph.dijkstra(v1,n,n))

if answer == float("inf"):
    print(-1)
else:
    print(answer)